Technical Onboarding - zirah

ziriraha opened the pull request.

ziriraha committed:
[ADD] estate: add an application for managing real estate

The estate app allows the managing of real estate properties, types of properties, tags and offers.
Simplifies the pipeline of creating a property, receiving offers and finally selling the property.
There is the possibility of creating an invoice automatically on property sold.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/939.png)](https://mergebot.odoo.com/odoo/tutorials/pull/939)

alialfie reviewed (commented):
Nice job! Few small comments :D

alialfie commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,43 @@
+
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date("Available from", copy=False, default=fields.Date.add(fields.Date.today(), months=3))
```
```suggestion
    date_availability = fields.Date(
        "Available from",
        copy=False,
        default=fields.Date.add(fields.Date.today(), months=3),
    )
```

alialfie commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+
+    @api.depends('living_area', 'garden_area')
+    def _compute_total_area(self):
+        for record in self:
```
It's better to use another name than `record`. Something like `property` in your case

alialfie commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+
+    @api.depends('offer_ids.price')
+    def _compute_best_price(self):
+        for record in self:
```
idem

alialfie commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+            self.garden_orientation = None
+
+    def action_set_sold(self):
+        for record in self:
```
idem

alialfie commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+        return True
+
+    def action_set_cancelled(self):
+        for record in self:
```
idem

alialfie commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,113 @@
+
+    @api.constrains('selling_price', 'expected_price')
+    def _check_price_difference(self):
+        for record in self:
```
idem

alialfie commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,75 @@
+
+    @api.depends('validity')
+    def _compute_date_deadline(self):
+        for record in self:
```
idem (I won't mention all of them actually, i trust you'll check them :D)

alialfie reviewed (commented):
Nice job! Few comments and should be good after :D

ziriraha committed:
[ADD] estate: test some functionalities before starting.

created 3 tests: ensure that a property cannot be sold without accepted offers, ensure that an offer cannot be created for a sold property, and ensure that selling a property sets the state to sold.

ziriraha committed:
[ADD] estate: add for garden checkbox in the form

ziriraha committed:
[ADD] estate: improve security by setting agent and manager roles.

ziriraha committed:
[ADD] estate: limit editing of a property to only assigned salesperson.

ziriraha committed:
[FIX] estate: correct if condition and permissions

ziriraha committed:
[ADD] estate: override invoice permission to be able to sell property

ziriraha committed:
[FIX] estate: add the status accepted to the offer of the test method.

ziriraha committed:
[ADD] estate: check security level before trying to create invoice.

ziriraha committed:
[ADD] estate: restrict permissions of properties by company.

ziriraha committed:
[ADD] estate: restrict access to menu to managers only

alialfie reviewed (commented):
Very noice!

ziriraha committed:
[REF] estate: change property state to be a computed field.

Until now the state of a property is handled with manual assignments when it could be easily computed by in which state the property is in. This reduces complexity when adding new features on the matter of having to handle the state manually.
The change includes the creation of two variables (which should not be used outside the estate property file) to handle if a property is sold or cancelled.

ziriraha commented on estate_account/models/estate_property.py:
```diff
@@ -34,6 +34,6 @@ def action_set_sold(self):
             invoice_vals_list.append(invoice_vals)
 
-        self.env['account.move'].create(invoice_vals_list)
+        self.env['account.move'].sudo().create(invoice_vals_list)
```
A real estate agent might not have the permission to create invoices or access the invoice app, the sudo() is needed so they are still able to sell a property (and create a draft invoice).

ziriraha committed:
[LINT] estate: change super to conform with ci/style.

ziriraha committed:
[REF] estate: remove complexity by allowing the state to be set manually too.
